### ISSUE TITLE
Support ESM applications and configs

### ISSUE THREAD

**Issue by eng-dibo** [just-jeb/angular-builders#1101]
this package cannot be used with es modules (i.e packages that have `type: module`)

in [this line](https://github.com/just-jeb/angular-builders/blob/74f3bc99e15a78a9e1bf669b3cb553ed77756aa1/packages/custom-webpack/src/custom-webpack-builder.ts#L75) you are using `require()` to load the custom webpack config file. if custom.webpack.config.ts is inside an es module, you must use `import()`

```
// should be replaced with `import(path)`
  const customWebpackConfig = require(path);
```

also it has to handle webpack.config that written in typescript i.e `custom.webpack.ts`
you will need to pass `--experimental-specifier-resolution=node` to ts-node, so it can work with `.ts`

I use Angular v13 and custom-webpack v13

[minimal reproduction](https://stackblitz.com/edit/node-mdn6pg)
run `ng build`

**Comment by just-jeb** [comment]
The issue is a bit deeper than just one `require`, we have an open issue (#978) for this. Any help is welcome 😃 
As for TS, it is supported and is mentioned in the [docs](https://github.com/just-jeb/angular-builders/tree/master/packages/custom-webpack#custom-webpack-config-object).

**Comment by eng-dibo** [comment]
> As for TS, it is supported and is mentioned in the [docs](https://github.com/just-jeb/angular-builders/tree/master/packages/custom-webpack#custom-webpack-config-object).

I saw it, I mean with something like `node --loader ts-node/esm` you cannot use `.ts` unless you pass the flag `--experimental-specifier-resolution=node`

**Comment by just-jeb** [comment]
I'm not sure I understand your use case. Could you elaborate on functionality you'd like to see here?

**Comment by eng-dibo** [comment]
try to run this command `node --loader ts-node/esm file.ts`. nodejs will complain about the unknown extension `.ts`
you need to provide the flag `--experimental-specifier-resolution=node`

in our case, it will complain about `webpack.ts`. since we build Angular apps using `ng` instead of `node` I don't know how to pass this flag

**Comment by eng-dibo** [comment]
I prepared a [minimal reproduction](https://stackblitz.com/edit/node-mdn6pg)

**Comment by just-jeb** [comment]
It is possible that I'm [doing](https://github.com/just-jeb/angular-builders/pull/1114/files) something wrong, however it seems that once I fix the builder itself the architect causes another trouble: 
```console
An unhandled exception occurred: Must use import to load ES Module: /Users/jeb/angular-builders/packages/custom-webpack/dist/browser/index.js
require() of ES modules is not supported.
require() of /Users/jeb/angular-builders/packages/custom-webpack/dist/browser/index.js from /Users/jeb/angular-builders/node_modules/@angular-devkit/architect/node/node-modules-architect-host.js is an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which defines all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use import(), or remove "type": "module" from /Users/jeb/angular-builders/packages/custom-webpack/package.json.
``` 

Now, this error is coming from [here](https://github.com/angular/angular-cli/blob/master/packages/angular_devkit/architect/node/node-modules-architect-host.ts#L200), which does use the `await import` syntax, but apparently it is compiled to this (from my `node_modules`, architect version `0.1301.3`):
```js
async loadBuilder(info) {
    const builder = (await Promise.resolve().then(() => __importStar(require(info.import)))).default;
    if (builder[internal_1.BuilderSymbol]) {
        return builder;
    }
    // Default handling code is for old builders that incorrectly export `default` with non-ESM module
    if (builder === null || builder === void 0 ? void 0 : builder.default[internal_1.BuilderSymbol]) {
        return builder.default;
    }
    throw new Error('Builder is not a builder');
}
```

If you want to see it for yourself you're welcome to check out the [branch](https://github.com/just-jeb/angular-builders/tree/esm-support): 
1. `yarn --frozen-lockfile`
2. `yarn build:packages`
3. `cd examples/custom-webpack/sanity-app`
4. `yarn build -c mjs`

**Comment by just-jeb** [comment]
Btw #978 is the same issue.

**Comment by eng-dibo** [comment]
> Now, this error is coming from [here](https://github.com/angular/angular-cli/blob/master/packages/angular_devkit/architect/node/node-modules-architect-host.ts#L200), which does use the `await import` syntax

exactly

**Comment by eng-dibo** [comment]
good effort <3

**Comment by eng-dibo** [comment]
you can get some inspiration from [this](https://github.com/angular/angular-cli/tree/master/packages/angular_devkit/build_webpack)

**Comment by just-jeb** [comment]
The PR is insightful, thanks.  
I was originally thinking of something like this (i.e. if you're in ESM context use dynamic `import`, otherwise `require`).   
However, while it is possible to check for config file extension, it's quite tricky to identify the scope you're in. I see that @alan-agius4 is dealing with this by wrapping `require` with `try` and `catch`, not sure about using exceptions in happy flows 😅 .  
Another option I was looking into is having a [hybrid](https://2ality.com/2019/10/hybrid-npm-packages.html) package that supports both ESM and CommonJs flows by having separate compilations for each.  
Haven't had enough time to research this though.

**Comment by eng-dibo** [comment]
you can check for esm by these steps
1- check file extension (.mjs for esm, .cjs for commonjs)
2- check the property `type: module` in the nearest package.json

also, you can define [conditional exports](https://nodejs.org/docs/latest-v16.x/api/packages.html#conditional-exports) in **your** package.json to deal with both cases

```
{

  "exports": {
    "import": "./main-module.mjs",
    "require": "./main-require.cjs"
  },
  "type": "module"
}
```

alternatively, you can go for @alan-agius4 's approach by just using try and catch blocks.

**Comment by just-jeb** [comment]
Yeah, conditional exports is what I meant by having to compilation targets.  
Regarding checking the `type` in nearest `package.json` I don't think it's a viable solution. Package managers do not promise specific node modules structure, so this builder can be under an Angular app, under one of it dependencies or even under a workspace root (in case of Yarn monorepo for example). Too fragile. 
As for the flag in config - I'd really like to avoid that but I might have no choice in regards of `ts-node`. We shall see. 
I think the `try catch` can be a temporary solution to unblock you and everyone else who's using ESM but the long term solution should rely on conditional exports.

**Comment by eng-dibo** [comment]
> Package managers do not promise specific node modules structure, so this builder can be under an Angular app, under one of it dependencies or even under a workspace root

in all cases nodejs only considers the nearest package.json. it may be found at the Angular directory level or up (each folder contains a package.json is a independent nodejs project and has its own type)

**Comment by just-jeb** [comment]
Actually `ts-node` loader that loads the `.ts` config uses the `tsconfig.app.json` as for today.   
So if add a `ts-node` section there you can [pass options](https://typestrong.org/ts-node/docs/configuration/#via-tsconfigjson-recommended) to `ts-node`.

**Comment by eng-dibo** [comment]
> `custom-webpack.config.ts` uses the `tsconfig.app.json` as for today.

I don't use `tsconfig.app.json` at all, in addition most people use different tsconfig files for the browser and the app.

that's why its better to add an option for the  custom tsconfig path

**Comment by just-jeb** [comment]
Well, technically it doesn't have to be `tsconfig.app.json` of course, I was referring to `architect.build.options.tsConfig` field in `angular.json`. You don't use it? How Angular build chain is configured in this case? 

> `ts-node` already has this option

Yeah, I believe this is what already [used](https://github.com/just-jeb/angular-builders/blob/master/packages/custom-webpack/src/utils.ts#L21) in this builder, it's just that the path to config is being taken from the general `tsConfig` build option. The reason is that I'm trying to keep the API surface minimal. However, if you have a use case for it, this decision can always be revisited. 

*Side note: Btw, ESM `indexTransform` will also have to be handled along with webpack config*

**Comment by eng-dibo** [comment]
> architect.build.options.tsConfig

yeah yeah, we don't need the suggested option so.

**Comment by eng-dibo** [comment]
> Well, technically it doesn't have to be `tsconfig.app.json` of course, I was referring to `architect.build.options.tsConfig` field in `angular.json`. You don't use it? How Angular build chain is configured in this case?
> 
> > `ts-node` already has this option
> 
> Yeah, I believe this is what already [used](https://github.com/just-jeb/angular-builders/blob/master/packages/custom-webpack/src/utils.ts#L21) in this builder, it's just that the path to config is being taken from the general `tsConfig` build option. The reason is that I'm trying to keep the API surface minimal. However, if you have a use case for it, this decision can always be revisited.
> 
> _Side note: Btw, ESM `indexTransform` will also have to be handled along with webpack config_



> ESM `indexTransform` will also have to be handled along with webpack config

I don't know why we need to add this feature to custom-webpack

the package must be modular and has only one specific feature

this reduces the chances that make custom-webpack incompatible with the original angular builder. there are many other solutions that offer this feature

**Comment by just-jeb** [comment]
The reasons are mostly historical. 
Originally `index.html` was generated as part of the webpack build and at some point it has been separated into a stand alone process that runs after the webpack build, but still as part of the `build-angular` builders.  
Since it's performed in the same builder in Angular CLI, you can't really separate this functionality into a separate builder.  
Maybe if the switch it to multiple aggregated builders later on we'll revise it again.  
  
I like to think of it as a repurposing of `custom-webpack` builder. In this context `custom-webpack` builder is not just a Custom Webpack builder, it's more like `extended-angular-builder`, which has the functionality of `webpack` config modification and `indexTransform`. 

Although semantics is important I don't think it's a good enough reason to rename the package 😅

**Comment by eng-dibo** [comment]
hope to see a new version as soon as possible, many projects deeply depend on custom-webpack

**Comment by just-jeb** [comment]
Planning to release it during the coming weekend. 

Btw, would you mind to share a bit more about your use case? I mean, Angular apps are using TypeScript, why would you need `"type": "module"` in an Angular app? Just curious.

**Comment by eng-dibo** [comment]
> why would you need `"type": "module"` in an Angular app? Just curious.

1. your project may include other stuff than Angular. Angular itself doesn't force switching to esm, but other situations may be involved.
2. you may depend on esm packages that don't compatible with commonjs (using condition exports for example)
3. you may need to get benefits from migrating to esm, like using `imports` to enable path aliasing natively

**Comment by just-jeb** [comment]
@eng-dibo Please try out `13.1.0-beta.0` (`@angular-builders/custom-webpack@next`).  
If all good I'll graduate it to stable.

**Comment by eng-dibo** [comment]
> @eng-dibo Please try out `13.1.0-beta.0` (`@angular-builders/custom-webpack@next`). If all good I'll graduate it to stable.

very excited to try it out.

**Comment by eng-dibo** [comment]
tested with `.ts` got this error `Unknown file extension ".ts" for webpack.config.ts`
solution:
pass `--experimental-specifier-resolution=node` to `node` or `ts-node`

example:
`node --experimental-modules --experimental-specifier-resolution=node -r tsconfig-paths/register --loader ts-node/esm ./webpack.config.ts`

`tsconfig-paths/register` doesn't work with esm

**Comment by just-jeb** [comment]
From the [docs](https://github.com/just-jeb/angular-builders/tree/master/packages/custom-webpack#es-modules-esm-support):  
> If you want to use TS config in ESM app, you must set the loader to ts-node/esm when running ng build. Also, in that case tsconfig.json for ts-node no longer defaults to tsConfig from the browser target - you have to specify it manually via environment variable. Example.
Note that tsconfig paths are not supported in TS configs within ESM apps. That is because tsconfig-paths do not support ESM.

**Comment by eng-dibo** [comment]
sorry, I wanted to mention providing the option `experimental-specifier-resolution=node` by the consumer when they want to use webpack.config in typescript

**Comment by just-jeb** [comment]
I think it is necessary only if you don't specify the file extension, right? For example [here](https://github.com/just-jeb/angular-builders/blob/master/examples/custom-webpack/sanity-app-esm/custom-webpack.config.ts#L4) it is unnecessary.  
Apart from this (which might or might not be part of the docs) is everything working as expected?

**Comment by eng-dibo** [comment]
initial tests for the `@next` version are doing well until now for many scenarios, testing may take a bit long time with me because my project's codebase is going a bit complicated.

thanks for your great efforts

**Comment by just-jeb** [comment]
No problem, I released it to a stable version `13.1.0`, lmk if there's any trouble.

**Comment by Parth7030** [comment]
+1 , still it is an issue with V13.3+

Any solution?

<img width="567" alt="image" src="https://github.com/just-jeb/angular-builders/assets/40611805/3a95e737-f1c1-4c94-b015-3e8f25c38026">
